spring-cloud-gcp-pubsub/src/main/java/org/springframework/cloud/gcp/pubsub/integration/inbound/PubSubMessageSource.java

```
@@ -0,0 +1,125 @@
+ * implementation of {@link AbstractMessageSource}.
+ *
+ * Enables more even load balancing with polling adapters, at the cost of efficiency gained from batched pulls.
+ *
+ * @author Elena Felder
+ *
+ * @since 1.1
+ */
+public class PubSubMessageSource extends AbstractMessageSource<Object> {
+
+	private static final Log LOGGER = LogFactory.getLog(PubSubMessageSource.class);
+
+	private final String subscriptionName;
+
+	private final PubSubSubscriberOperations pubSubSubscriberOperations;
+
+	private AckMode ackMode = AckMode.MANUAL;
+
+	private HeaderMapper<Map<String, String>> headerMapper = new PubSubHeaderMapper();
+
+	private Class payloadType = byte[].class;
+
+	public PubSubMessageSource(PubSubSubscriberOperations pubSubSubscriberOperations, String subscriptionName) {
+		Assert.notNull(pubSubSubscriberOperations, "Pub/Sub subscriber template can't be null.");
+		Assert.notNull(subscriptionName, "Pub/Sub subscription name can't be null.");
+		this.pubSubSubscriberOperations = pubSubSubscriberOperations;
+		this.subscriptionName = subscriptionName;
+	}
+
+	public AckMode getAckMode() {
+		return this.ackMode;
+	}
+
+	public void setAckMode(AckMode ackMode) {
+		this.ackMode = ackMode;
+	}
+
+	public Class getPayloadType() {
+		return this.payloadType;
+	}
+
+	public void setPayloadType(Class payloadType) {
+		this.payloadType = payloadType;
+	}
+
+	public HeaderMapper<Map<String, String>> getHeaderMapper() {
+		return this.headerMapper;
+	}
+
+	public void setHeaderMapper(HeaderMapper<Map<String, String>> headerMapper) {
+		this.headerMapper = headerMapper;
+	}
+
+	@Override
+	protected Object doReceive() {
+		List<ConvertedAcknowledgeablePubsubMessage> messages
+				= this.pubSubSubscriberOperations.pullAndConvert(this.subscriptionName, 1, true, this.payloadType);
+
+		if (messages.size() < 1) {
+			return null;
+		}
+
+		if (messages.size() > 1) {
+			LOGGER.warn("One message expected; " + messages.size() + " received. Returning the first one.");
```

Maybe throw an exception instead?

---

If it's a "should never happen" scenario, then it should be an exception IMO.